subsys/bluetooth/host/l2cap.c

```diff
@@ -724,13 +724,54 @@ static void l2cap_chan_rx_init(struct bt_l2cap_le_chan *chan)
 	}
 }
 
+static struct net_buf *l2cap_chan_le_get_tx_buf(struct bt_l2cap_le_chan *ch)
+{
+	struct net_buf *buf;
+
+	/* Return current buffer */
+	if (ch->tx_buf) {
+		buf = ch->tx_buf;
+		ch->tx_buf = NULL;
+		return buf;
+	}
+
+	return net_buf_get(&ch->tx_queue, K_NO_WAIT);
+}
+
+static int l2cap_chan_le_send_sdu(struct bt_l2cap_le_chan *ch,
+				  struct net_buf **buf, u16_t sent);
+
+static void l2cap_chan_tx_process(struct k_work *work)
+{
+	struct bt_l2cap_le_chan *ch;
+	struct net_buf *buf;
+
+	ch = CONTAINER_OF(work, struct bt_l2cap_le_chan, tx_work);
+
+	/* Resume tx in case there are buffers in the queue */
+	while ((buf = l2cap_chan_le_get_tx_buf(ch))) {
+		int sent = data_sent(buf)->len;
+
+		BT_DBG("buf %p sent %u", buf, sent);
+
+		sent = l2cap_chan_le_send_sdu(ch, &buf, sent);
+		if (sent < 0) {
+			if (sent == -EAGAIN) {
+				ch->tx_buf = buf;
+			}
+			break;
```

Comment:
bt_conn_send_cb would unref it if it fails to send.

Reply:
-EAGAIN is return whenever a segment cannot be send immediately which can happen under the following circuntances:

1. There are no credits
2. There are no buffers
3. There are no TX contexts

In all these case we just push the buffer back to the tx_queue to be send later. Note that what is dropped on number 3 is the segment buffer, the original buffer is unfacted since we restore the data if that cannot be sent.

Reply:
Thank you.
Me question about EMSGSIZE still remains then. I think in the case where `bt_l2cap_chan_send` submits the work, and  `l2cap_chan_tx_process` calls `l2cap_chan_le_send_sdu` and returns -EMSGSIZE the buffer seems to be not released, because the application would not receive this error code. It appears to me that the application has acquired the first ref but it is unclear who has the responsebility to release it.
This check here:
```
 	if (total_len > ch->tx.mtu) {
		return -EMSGSIZE;
	}
```
Maybe this check could be moved out of `l2cap_chan_le_send_sdu` in that case?

I'm looking at avdtp BTW, and it appears to not release the buf in case of errors returned. I think we need to document in the API of `bt_l2cap_chan_send` how error handling and buffer releasing should be handled.

Reply:
-EMSGSIZE will be generated only in the context of l2cap_chan_le_send_sdu since after that the buffer is no longer under the control of caller. I haven't changed the semantics of bt_l2cap_chan_send so Id leave any changes such as unrefing the buffer on error for later, and then yes we can update the documentation and do some of the necessary clean ups.